### PR TITLE
Introduce ROOT patches to clang 16

### PR DESCRIPTION
**Pull request by vepadulano**
ROOT has recently rebased cling on top of LLVM16, hence it requires a new build of clang on conda-forge with a new variant.

~~For now, there is no ROOT released version to pinpoint the variant, so I have used the tag `root_master`. The idea is to make the package available in our conda nightly builds, then it will be merged into the conda-forge releases at the next ROOT release (6.32).~~

ROOT 6.32 was released and the PR is now ready.

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/patches/generate_root_patches.py
Comment (h-vetinari) [inline recipe/patches/generate_root_patches.py:72]: Please add `--no-signature` here

Comment (vepadulano) [reply recipe/patches/generate_root_patches.py]: Done :+1:

**Comment by erykoff** [comment]
I wanted to highlight this issue (#265) to make sure that any new clang=16 root builds do _not_ have priority over default.  Thanks!

**Comment by h-vetinari** [comment]
> I wanted to highlight this issue (#265) to make sure that any new clang=16 root builds do _not_ have priority over default. Thanks!

This was taken care of for 16.x in https://github.com/conda-forge/clangdev-feedstock/pull/268. I tried to merge `16.x` here, but don't seem to have the rights to push to the remote, even as a maintainer (can't remember that happening before; the checkbox for allowing edits also enabled...). In any case, I'll leave that up to @vepadulano

**Comment by h-vetinari** [comment]
Any news here @vepadulano @chrisburr? :)

**Comment by vepadulano** [comment]
@h-vetinari thanks for the reminder! If it's not a problem for you, I will wait the official release of the next ROOT version (6.32) so that the clangdev variant can point to the correct LLVM tag from our repo. This should happen within the next month. Maybe this PR will need to be updated meanwhile, I will take care of that ofc

**Comment by vepadulano** [comment]
@h-vetinari The PR is ready for review now and points to the 6.32 tag just released.

**Comment by h-vetinari** [comment]
> @h-vetinari The PR is ready for review now and points to the 6.32 tag just released.

Great timing! I was thinking about reaching out because it seems like we'll have to drop support for clang <15 in the next libcxx release, and that would cut off the current root releases.

I tried to push some fixes that are necessary here before merging, but it doesn't work. Not sure why. I'll open a PR to your fork.

**Review by h-vetinari — APPROVED** [review]
I cannot review the root patches, but I trust you on this. The rest looks good. 👍 

@chrisburr, do you want to take a look as well?

**Comment by vepadulano** [comment]
@h-vetinari Thank you for your prompt reaction and the help, much appreciated! 

I have a question just for curiosity 

> we'll have to drop support for clang <15 in the next libcxx release

Does it mean that if we release a patch release of an older ROOT version that uses say clang 13 I will not be able to push changes to the `13.x` branch of this repo and thus publish  a new build for clang 13?

**Comment by h-vetinari** [comment]
> Does it mean that if we release a patch release of an older ROOT version that uses say clang 13 I will not be able to push changes to the `13.x` branch of this repo and thus publish a new build for clang 13?

You can still publish, but the problem is that once libcxx 18 is published, it'll start getting pulled in by our C++ compilers (as well as become a lower-bound for all contemporary packages through the run-export), so this has historically been an issue.

We try to extend clang version support of libcxx way beyond what upstream does (e.g. libcxx 17 only supports clang>=15, we've [tried](https://github.com/conda-forge/libcxx-testing-feedstock) to keep basic compatibility - modulo bugs due to old clang versions - all the way back to clang 11), but there are limits to how far we can push things. 😅

**Review by chrisburr — APPROVED** [review]
Thanks @h-vetinari for reviewing and keeping track of this! 🍰

**Pull request merged**
